vm/devices/storage/disk_layered/src/lib.rs

```diff
@@ -153,23 +148,45 @@ impl LayeredDisk {
     ///
     /// The layers must be ordered from top to bottom, with the top layer being
     /// the first in the list.
-    pub fn new(
+    pub async fn new(
         read_only: bool,
-        mut layers: Vec<LayerConfiguration>,
+        layers: Vec<LayerConfiguration>,
     ) -> Result<Self, InvalidLayeredDisk> {
         if layers.is_empty() {
             return Err(InvalidLayeredDisk::NoLayers);
         }
 
+        let mut attached_layers: Vec<LayerConfiguration<AttachedDiskLayer>> = Vec::new();
         // Collect the common properties of the layers.
         let mut last_write_through = true;
         let mut is_fua_respected = true;
         let mut optimal_unmap_sectors = 1;
         let mut unmap_must_zero = false;
         let mut disk_id = None;
         let mut unmap_behavior = UnmapBehavior::Zeroes;
-        for (i, config) in layers.iter().enumerate() {
+
+        let mut lower_layer_metadata = None;
+        for (i, config) in layers.into_iter().enumerate() {
             let layer_error = |e| InvalidLayeredDisk::Layer(i, e);
+
+            let config = {
+                attached_layers.push(LayerConfiguration {
```

Comment:
Don't we need to do the attach in reverse order?

Reply:
Yeah, layer zero is the top layer.

Reply:
Ok, should be fixed now. 

Admittedly, I took a pretty simply approach to fixing this, where I split the attach loop into a distinct step prior to validating the layer stack. I did this as a quick way to get some test runs in... but I suspect this isn't the optimal approach?